dartagnan/src/main/java/com/dat3m/dartagnan/analysis/graphRefinement/graphs/GraphHierarchy.java

```diff
@@ -0,0 +1,213 @@
+package com.dat3m.dartagnan.analysis.graphRefinement.graphs;
+
+import com.dat3m.dartagnan.analysis.graphRefinement.graphs.eventGraph.EventGraph;
+import com.dat3m.dartagnan.analysis.graphRefinement.graphs.eventGraph.GraphListener;
+import com.dat3m.dartagnan.analysis.graphRefinement.graphs.eventGraph.unary.RecursiveGraph;
+import com.dat3m.dartagnan.utils.dependable.DependencyGraph;
+import com.dat3m.dartagnan.verification.model.Edge;
+import com.dat3m.dartagnan.verification.model.ExecutionModel;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+//TODO: Maybe add a concrete way of backtracking
+
+// This class should manage a hierarchy of event graphs such that the following operations are supported:
+// (1) Adding edges and propagating changes
+// (2) Backtracking to a previous state
+// (3) Listening to changes in graphs
+
+public class GraphHierarchy {
+    private final Map<EventGraph, Set<GraphListener>> graphListenersMap;
+    private final DependencyGraph<EventGraph> dependencyGraph;
+    private final PriorityQueue<Task> tasks = new PriorityQueue<>();
+
+    // ============== Construction ===============
+
+    public GraphHierarchy(Set<EventGraph> eventGraphs) {
+        dependencyGraph = DependencyGraph.from(eventGraphs);
+        graphListenersMap = new HashMap<>(dependencyGraph.getNodeContents().size() * 4 / 3);
+        dependencyGraph.getNodeContents().forEach(x -> graphListenersMap.put(x, new HashSet<>()));
+    }
+
+    // ========================================
+
+    // ============= Accessors =================
+
+    public Set<EventGraph> getGraphs() { return graphListenersMap.keySet(); }
+
+    // The list of graphs in topological order
+    public List<EventGraph> getGraphList() { return dependencyGraph.getNodeContents(); }
+
+    public EventGraph getRootGraph() { return getGraphList().get(0); }
```

Comment:
Unless the WMM has not a single relation, this should never be empty. And since every of our WMMs has at least rf, I think this should never be empty. But I will double check that.